gradle-incremental-configuration-cache/src/main/java/com/palantir/gradle/configcache/incremental/AbstractRunTask.java

```diff
@@ -36,44 +35,79 @@
+public abstract class AbstractRunTask extends DefaultTask {
 
     @InputFiles
     @PathSensitive(PathSensitivity.RELATIVE)
-    public abstract RegularFileProperty getDryRunTasksFile();
+    public abstract RegularFileProperty getTasksToRunFile();
 
     @Input
     @org.gradle.api.tasks.Optional
     public abstract Property<String> getInitScript();
 
+    @Input
+    public abstract Property<Boolean> getUseClonedDirectory();
+
     @Inject
     protected abstract ProjectLayout getProjectLayout();
 
+    @Inject
+    protected abstract ExecOperations getExecOperations();
+
     @OutputFile
     public abstract RegularFileProperty getMarkerOutputFile();
 
-    public AbstractDryRunTask() {
+    public AbstractRunTask() {
         getMarkerOutputFile()
                 .set(getTemporaryDir().toPath().resolve(getName() + ".marker").toFile());
+        getUseClonedDirectory().convention(false);
     }
 
-    protected final DryRunResult dryRun(List<String> extraArgs) throws IOException {
+    protected final RunResult run(List<String> extraArgs) throws IOException {
         Set<String> tasksToDryRun =
-                new TaskListFile(getDryRunTasksFile().getAsFile().get().toPath()).loadTasks();
+                new TaskListFile(getTasksToRunFile().getAsFile().get().toPath()).loadTasks();
 
         if (tasksToDryRun.isEmpty()) {
-            getLogger().info("No tasks to dry-run");
-            return DryRunResult.success(new TreeSet<>());
+            getLogger().info("No tasks to run");
+            return RunResult.success("");
+        }
+
+        getLogger().info("Running {} tasks", tasksToDryRun.size());
+
+        return runTasksInDirectory(determineProjectDirectory(), tasksToDryRun, extraArgs);
+    }
+
+    private File determineProjectDirectory() throws IOException {
+        if (!getUseClonedDirectory().get()) {
+            return getProjectLayout().getProjectDirectory().getAsFile();
         }
 
-        getLogger().info("Dry-running {} tasks", tasksToDryRun.size());
+        Path cloneDir = Files.createTempDirectory("gradle-config-cache-validation-");
+        getLogger().info("Cloning repository to temporary directory: {}", cloneDir);
 
-        GradleConnector connector = GradleConnector.newConnector()
-                .forProjectDirectory(getProjectLayout().getProjectDirectory().getAsFile());
+        cloneRepository(getProjectLayout().getProjectDirectory().getAsFile(), cloneDir.toFile());
+
+        return cloneDir.toFile();
+    }
+
+    private void cloneRepository(File sourceDir, File targetDir) {
+        getExecOperations().exec(execSpec -> {
+            execSpec.setWorkingDir(targetDir.getParentFile());
+            execSpec.commandLine(
+                    "git", "clone", "--quiet", "--shared", sourceDir.getAbsolutePath(), targetDir.getName());
+            execSpec.setIgnoreExitValue(false);
+        });
+
+        getLogger().info("Successfully cloned repository to: {}", targetDir);
```

Comment:
Not sure log lines like this add much tbh. Either it fails with an exception or it succeeds.